File: src/linter.ts
import { CommitParser } from "conventional-commits-parser";
import type {
  ParsedCommit,
  RulesConfig,
  RuleConfigSeverity,
  RuleConfigCondition,
  LintOutcome,
  LintRuleOutcome,
  ParserOptions,
} from "./types.js";
import rulesRegistry from "./rules/index.js";

export function parseCommit(message: string, options?: ParserOptions): ParsedCommit {
  const parserOpts = {
    headerPattern: /^(\w*)(?:\((.*)\))?!?: (.*)$/,
    headerCorrespondence: ["type", "scope", "subject"],
    ...options,
  };
  const parser = new CommitParser(parserOpts as any);
  const parsed = parser.parse(message) as any;
  return {
    raw: message,
    header: parsed.header || message.split(/\r?\n/)[0] || null,
    type: parsed.type || null,
    scope: parsed.scope || null,
    subject: parsed.subject || null,
    body: parsed.body || null,
    footer: parsed.footer || null,
    notes: parsed.notes || [],
    references: parsed.references || [],
    revert: parsed.revert || null,
    mentions: parsed.mentions || [],
  };
}

export async function lintCommit(
  message: string,
  rulesConfig: RulesConfig = {},
  options: {
    parserOpts?: ParserOptions;
    defaultIgnores?: boolean;
    ignores?: Array<(msg: string) => boolean>;
  } = {},
): Promise<LintOutcome> {
  const trimmed = message.trim();

  // Handle ignores
  if (options.defaultIgnores !== false) {
    if (
      trimmed.startsWith("Fixup!") ||
      trimmed.startsWith("squash!") ||
      trimmed.startsWith("Merge ") ||
      trimmed.startsWith("Revert ")
    ) {
      return { input: message, valid: true, errors: [], warnings: [] };
    }
  }
  if (options.ignores && options.ignores.some((ignore) => ignore(message))) {
    return { input: message, valid: true, errors: [], warnings: [] };
  }

  const parsed = parseCommit(message, options.parserOpts);
  const errors: LintRuleOutcome[] = [];
  const warnings: LintRuleOutcome[] = [];

  for (const [ruleName, ruleConfig] of Object.entries(rulesConfig)) {
    if (!ruleConfig) continue;

    const level: RuleConfigSeverity = ruleConfig[0];
    if (level === 0) continue; // Disabled

    const when: RuleConfigCondition = ruleConfig[1] || "always";
    const value = ruleConfig[2];

    const ruleFn = rulesRegistry[ruleName];
    if (!ruleFn) {
      // Ignore unknown rule gracefully or flag warning
      continue;
    }

    try {
      const outcome = await ruleFn(parsed, when, value);
      const [valid, messageOverride] = outcome;
      if (!valid) {
        const item: LintRuleOutcome = {
          valid: false,
          level,
          name: ruleName,
          message: messageOverride || `Rule ${ruleName} failed`,
        };
        if (level === 2) {
          errors.push(item);
        } else if (level === 1) {
          warnings.push(item);
        }
      }
    } catch (err: any) {
      errors.push({
        valid: false,
        level: 2,
        name: ruleName,
        message: err.message || `Error executing rule ${ruleName}`,
      });
    }
  }

  return {
    input: message,
    valid: errors.length === 0,
    errors,
    warnings,
  };
}

export default lintCommit;

File: src/linter.rs
use crate::parser::parse_commit;
use crate::rules::evaluate_rule;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LintRuleOutcome {
    pub valid: bool,
    pub level: u8,
    pub name: String,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LintOutcome {
    pub input: String,
    pub valid: bool,
    pub errors: Vec<LintRuleOutcome>,
    pub warnings: Vec<LintRuleOutcome>,
}

pub fn lint_commit(message: &str, rules_config: &HashMap<String, Value>) -> LintOutcome {
    let trimmed = message.trim();

    if trimmed.starts_with("Fixup!")
        || trimmed.starts_with("squash!")
        || trimmed.starts_with("Merge ")
        || trimmed.starts_with("Revert ")
    {
        return LintOutcome {
            input: message.to_string(),
            valid: true,
            errors: Vec::new(),
            warnings: Vec::new(),
        };
    }

    let parsed = parse_commit(message);
    let mut errors = Vec::new();
    let mut warnings = Vec::new();

    for (rule_name, rule_value) in rules_config {
        if let Some(arr) = rule_value.as_array() {
            if arr.is_empty() {
                continue;
            }

            let level = arr.first().and_then(|v| v.as_u64()).unwrap_or(0) as u8;
            if level == 0 {
                continue;
            }

            let when = arr.get(1).and_then(|v| v.as_str()).unwrap_or("always");
            let value = arr.get(2).unwrap_or(&Value::Null);

            let (valid, message_override) = evaluate_rule(rule_name, &parsed, when, value);

            if !valid {
                let outcome = LintRuleOutcome {
                    valid: false,
                    level,
                    name: rule_name.clone(),
                    message: if message_override.is_empty() {
                        format!("Rule {} failed", rule_name)
                    } else {
                        message_override
                    },
                };
                if level == 2 {
                    errors.push(outcome);
                } else if level == 1 {
                    warnings.push(outcome);
                }
            }
        }
    }

    LintOutcome {
        input: message.to_string(),
        valid: errors.is_empty(),
        errors,
        warnings,
    }
}
